newlib/libc/machine/riscv/memcpy.S

```diff
@@ -0,0 +1,247 @@
+/* Copyright (c) 2019  SiFive Inc. All rights reserved.
+   Copyright (c) 2025 Mahmoud Abumandour <ma.mandourr@gmail.com>
+
+   This copyrighted material is made available to anyone wishing to use,
+   modify, copy, or redistribute it subject to the terms and conditions
+   of the FreeBSD License.   This program is distributed in the hope that
+   it will be useful, but WITHOUT ANY WARRANTY expressed or implied,
+   including the implied warranties of MERCHANTABILITY or FITNESS FOR
+   A PARTICULAR PURPOSE.  A copy of this license is available at
+   http://www.opensource.org/licenses.
+*/
+
+#include <sys/asm.h>
+
+
+
+#ifdef __riscv_zbkb
+
+    #if __riscv_xlen == 32
+        /* 32-bit + zbkb */
+        #define PACK4(r1, r2, r3, r4) \
+            packh r1, r1, r2;         \
+            packh r3, r3, r4;         \
+            pack r1, r1, r3
+    #else
+        /* 64-bit + zbkb */
+        #define PACK4(r1, r2, r3, r4) \
+            packh r1, r1, r2;         \
+            packh r3, r3, r4;         \
+            packw r1, r1, r3
+    #endif
+#else
+    #define PACK4(a4, a5, s0, s1) \
+        slli  s1, s1, 24;        \
+        slli  s0, s0, 16;        \
+        slli  a5, a5, 8;         \
+        or    a4, a4, a5;        \
+        or    a4, a4, s0;        \
+        or    a4, a4, s1
+#endif
+
+
+.text
+.global memcpy
+.type	memcpy, @function
+memcpy:
+#if defined(PREFER_SIZE_OVER_SPEED) || defined(__OPTIMIZE_SIZE__)
+  mv a3, a0
+  beqz a2, 2f
+
+1:
+  lbu a4, 0(a1)
+  sb a4, 0(a3)
+  add   a2, a2, -1
+  add   a3, a3, 1
+  add   a1, a1, 1
+  bnez a2, 1b
+
+2:
+  ret
+
+#else
+    mv a3, a0
+    li    a4, SZREG
+    addi  sp, sp, -2 * SZREG
+    REG_S s0, 0(sp)
+    REG_S s1, SZREG(sp)
+    bleu  a2, a4, .Lcopy1
+
+#if defined(__riscv_misaligned_avoid) || defined(__riscv_misaligned_slow)
+    andi a4, a1, SZREG - 1
+    andi a5, a3, SZREG - 1
+    beq a4, a5, .Lalign
+    li s0, SZREG
+    sub s0, s0, a4
+    sub a2, a2, s0
+
+;# In the case where source and destination are not aligned, we align the
+;# destination, load SZREG bytes, one at a time, from source, and then store them
+;# to destination in one instruction.
+.Lalign_dest:
+    lbu a4, 0(a1)
+    sb  a4, 0(a3)
+    addi a1, a1, 1
+    addi a3, a3, 1
+    addi s0, s0, -1
+
+    bnez s0, .Lalign_dest
+
+.Ldest_aligned:
+    li s0, SZREG
+    bleu a2, s0, .Lcopy1
+
+    lbu a4, 0(a1)
+    lbu a5, 1(a1)
+    lbu s0, 2(a1)
+    lbu s1, 3(a1)
+
+    PACK4(a4, a5, s0, s1)
+#if __riscv_xlen == 64
+
+    lbu t0, 4(a1)
+    lbu a5, 5(a1)
+    lbu s0, 6(a1)
+    lbu s1, 7(a1)
+
+    PACK4(t0, a5, s0, s1)
+    pack a4, a4, t0
```

Comment:
i would move this inside the macro.
Also, I believe splat is the more fitting name. 
llvm libc has a similar library function:
https://github.com/llvm/llvm-project/blob/be6ccc98f38227db02164f17bfaf0ac86d800e4a/libc/src/string/memory_utils/op_generic.h#L129